src/transformStateWithClones.js

```diff
@@ -7,7 +7,48 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const clone = { ...state };
+  const stateHistory = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        stateHistory.push({ ...addProperties(clone, action.extraData) });
+        break;
+
+      case 'removeProperties':
+        stateHistory.push({ ...removeProperties(clone, action.keysToRemove) });
+        break;
+
+      case 'clear':
+        stateHistory.push({ ...clearProperties(clone) });
+        break;
+    }
```

Comment:
[CODE STYLE]: Use a default case in your switch statement for error handling to ensure that any unexpected action types are caught and handled appropriately.